### ISSUE TITLE
Molecular Graph Convolutions: Moving Beyond Fingerprints

### ISSUE THREAD

**Issue by agitter** [greenelab/deep-review#53]
https://arxiv.org/abs/1603.00856

**At a glance**: Related to virtual screening #45.  Similar to #52, provides an alternative molecular fingerprint for chemical compounds.  This paper uses graph convolutions in a neural network.

**Comment by cgreene** [comment]
@agitter This is awesome! Someone asked on twitter either late last year or early this year what I thought was an interesting method and I pitched convolution on graphs!

I'm sure it doesn't pre-date their work since it came out only a couple months later, but cool!
https://twitter.com/GreeneScientist/status/689863461879291904

**Comment by cgreene** [comment]
Also @davek44 quickly pointed out that it was already used:
https://twitter.com/drklly/status/689884899063566336
Jerk! ;)

**Comment by agitter** [comment]
This was something I had been interested in for a while so I too was excited to see this paper.  I think @davek44 is referring to this other graph paper #52.

**Comment by agitter** [comment]
Published this week  http://doi.org/10.1007/s10822-016-9938-8

**Comment by agitter** [comment]
### Biology
- Molecular graph has atoms as nodes and bonds as edges, which both have some input properties
- For deep learning, want a more raw feature representation based on the graph; typically molecular fingerprints are used in many machine learning applications (as in #55)
- Provides overview of common feature representations including extended-connectivity fingerprints (ECFP) and rapid overlay of chemical structures (ROCS)
- More related work on molecular graphs besides #52, may want to look at http://doi.org/10.1021/ci400187y
- The main focus is a new feature representation for molecules, but they test it using a virtual screening dataset similar to the one in #55

### Computational aspects
- Convolutions are for undirected graphs, cite some related work on convolution in networks
- Must operate on graphs of varying sizes; molecules have different numbers of nodes and edges
- Model properties: order invariance (atom and bond order in input), permutation invariance, pair order invariance (node symmetry in pair layers)
- Atom layers (2D) encode information about each of the *n* atoms (graph nodes)
- Pair layers (3D) encode information (such as shortest path graph distance) about each of the *n* X *n* pairs of atoms (graph edges)
- Convolution here means applying the same function to every atom in an atom layer, every pair in a pair layer, or subsets of pairs filtered by graph distance; used for atom-to-atom and pair-to-pair operations
- Pair-to-atom layer (edge-to-node) operation generates an output for each node *a* by applying a function to all edges incident to *a* in the pair layer and summing their outputs
- Atom-to-pair layer (node-to-edge) operation is similar; apply one or more functions that all take nodes *a* and *b* as input and sum their outputs
- Put these operations together and get a Weave module: atom layer and pair layer as input, apply various transformations described above, new atom layer and pair layer as output
- Can stack Weave modules like Inception modules for image analysis, optionally operating on local regions of the graph initially and then expanding to greater graph distances
- After applying the Weave modules they have a feature vector for each atom and each pair but still need a summary of the entire graph; create a fuzzy histogram for each dimension of the atoms' feature vectors; does not use the pairs directly
- Finally they have the features for the entire molecule and proceed with standard fully connected layers and softmax output
- The current application considers molecular graphs with at most 60 nodes and graph distance of at most 2 edges; uses two Weave modules and two fully connected layers with 2000 and 100 hidden units as a baseline but varies these during evaluation
- The input is 27 chemical features for atoms and 12 for atom pairs
- Evaluate using AUC and ROC enrichment as recommended [here](http://doi.org/10.1007/s10822-008-9196-5)
- Compare to multitask network from #55, random forest, and logistic regression that use fingerprint features as well as a naive baseline
- Neural networks beat the other classifiers; the multitask network in #55 is very similar to or slightly better than the graph convolutional network in most cases
- Try to replicate the evaluation in #52 and are slightly better on two tasks, significantly better on the third
- Evaluate consequences of changing the network structure and maximum graph distance
- No code

### Why include it in the review
- Graph convolutions and this particular design are fascinating from a computational perspective
- Biologically, this is a creative use of deep learning to learn new feature representations for small molecules
- Could perhaps have applications in other biological networks if code had been provided; I see nothing here that is specific to molecular graphs
- Molecular graphs are much smaller than most biological networks so other applications might involve pathway-size graphs as opposed to PPI network scale
- Beyond graphs, it is conceivable that their ideas for transforming the final atom-based features to molecule-level features could be applied to unordered single-cell data (e.g. #79)

**Comment by swamidass** [comment]
@agitter wrote: @swamidass I agree with the perspective that the field is in a "creative experimentation" phase. I'll have to go back to the paper, but my notes indicate that #53 did a good job comparing graph convolutions to circular fingerprints in a fair manner. 

==========

I reread the paper to be sure. Unfortunately, this is not correct; they did not do a good job comparing to fingerprints in a fair manner. They used an approach that is known to be poorly performing.  They basically used standard fingerprints as a binary vector input into a LR or NN. This is a known bad approach. 

They should have used Tanimoto or MinMax similarity (a "zero-shot approach" if you will). They also should have used AUC to separate actives from inactives (to model the actual use case). Doing this, circ fingerprints almost certainly would have outperformed them. And adding a neuralnetwork on top of this (as in IRVs) would have likely done even better.

Moreover (and this is extremely important), they only looked at 3 relatively small datasets of very different types, and did not compare to other work. Recall, a paper published that same year was validating against MANY state-of-the art methods on THOUSANDS of DATASET. There is just no comparison here. This paper did not compare to the state-of-the-art.

**Comment by swamidass** [comment]
I should add that they did compare their method to IRV (good!) on one dataset (bad). They were not able to outperform it. Remember, the IRV has about 10 weights, and trains in minutes on a laptop. They had perhaps millions of weights, and take hours to train.

**Comment by cgreene** [comment]
@swamidass / @agitter : This seems worth including - is it safe to say that in bio domains with a strong history of feature engineering & potentially limited examples that it can be tough for deep NNs to outperform existing methods?

**Comment by cgreene** [comment]
This being the discussion that you are having. :+1:

**Comment by swamidass** [comment]
That is exactly correct. And in these domains, using Deep Learning techniques to incorporate domain knowledge and reduce parameters can be very effective.

**Comment by cgreene** [comment]
Ok - yes - we should make sure to include that. Do you think there is a horizon (perhaps data quantity, perhaps something else) at which point any benefit from the history of feature engineering will be balanced by deep NN's ability to construct reasonable features given sufficient data (almost certainly at a cost of interpretability)? Or are our current models so strong that this is nearly impossible to imagine. I think that sharing *that* kind of perspective + an argument as to why is the primary goal of this review. We can be wrong - I'm comfortable with that - as long as we think the limb is reasonably well supported.

**Comment by swamidass** [comment]
In my experience (big caveat) in chemical modeling a basic rule still seems to hold. Keeping the number of parameters less than the number of instances in the most rare class usually leads to the most robust performance. Some modifications of that rule are appropriate; dropout of 0.5 can double the number of weights you use. And of course, one can get reasonable performance out of much higher parameters than examples (especially with modern techniques), but the simpler approaches that leverage the collective wisdom of the last several decades of chemical informatics (like Tanimoto SVMs or IRVs for chemical screening) seem to currently work quantitatively better. For example, IRVs can work in cases where you just have 10 or 15 positive examples, but even SVMs can fail when data is this scarce.

I think the real advantage of neural networks is when they allow you to seamlessly use more types of data than you could otherwise incorporate, or if they can give us specific insight into the underlying behavior of the system.

If we see the cases in chemistry where truly deep networks have demonstrated superior performance (e.g. reaction prediction in Baldi's work and metabolism/reactivity prediction in my work, SimulationsPlus, and Optibrium), these cases have much better balanced training data and much more of it (at the level that is presented to the network). In this context, they truly can excell. However, best performance still requires carefully constructed features.

One striking example of this is aromaticity. It appears that none of the current deep learning architectures in chemistry can efficiently represent the rules of aromaticity, which is of course a fundamentally important feature of molecules. Both Baldi and our groups handle this by adding this in as a precomputed feature to our networks, and this does strongly influence predictions. This does not seem like a permanent problem, as better architectures could solve this problem (and the large class of similar cases), but we have not yet arrived at that point yet.

**Comment by agitter** [comment]
Regarding this paper and IRV, I agree that a more powerful comparison would have compared the graph convolution model to IRV on many more datasets, for example, the 100+ they used in Table 4.  I have no idea how hard it would have been to add IRV as a baseline in the PCBA, MUV, and Tox21 testing.  But I applaud the authors for comparing to your published results and being willing to admit that they didn't beat them.

I'll post more in #174 about the area in general.
